### ISSUE TITLE
Single file application publishing causes an exception to be thrown

### ISSUE THREAD

**Issue by dnlib** [Kaliumhexacyanoferrat/GenHTTP#268]
When starting the GenHTTP webserver with a single-file deployment or embedding the webserver DLLs, GenHTTP fails to start with the following exception. Instead of actually starting up correctly.

I believe this issue is to do with the `assembly.Location` call which is bugged if the app has been deployed as single-file.

```
System.ArgumentException: The path is empty. (Parameter 'path')
   at System.IO.Path.GetFullPath(String path)
   at System.IO.FileInfo..ctor(String originalPath, String fullPath, String fileName, Boolean isNormalized)
   at System.IO.FileInfo..ctor(String fileName)
   at GenHTTP.Modules.IO.Embedded.EmbeddedResourceBuilder.Build()
   at GenHTTP.Modules.Placeholders.Placeholders.Template[T](IBuilder`1 templateProvider)
   at GenHTTP.Engine.Infrastructure.CoreRouter..ctor(IHandlerBuilder content, IEnumerable`1 concerns, Boolean development)
   at GenHTTP.Engine.Infrastructure.ThreadedServerBuilder.Build()
   at GenHTTP.Engine.Hosting.ServerHost.Build()
   at GenHTTP.Engine.Hosting.ServerHost.Start()
   at GenHTTP.Engine.Hosting.ServerHost.Run()
```

**Example**

```csharp
        var api = Layout.Create()
            .AddService<TestProvider>("test")
            .Add(CorsPolicy.Permissive());

        var int1 = Host.Create()
            .Handler(api)
            .Defaults(true, false, false, false, true)
            .RequestReadTimeout(TimeSpan.FromSeconds(5))
            .Compression(new CompressionConcernBuilder()
                .Add(new BrotliCompression())
                .Add(new GzipAlgorithm()))
            .Backlog(3)
#if DEBUG
            .Development()
#endif
            .Run();
```

**Additional Information**

- App was published using single-file deployment, linux-x64 runtime
- Latest version from NuGet
- Running on Ubuntu 20.04.3 LTS (server) (hyper-v)

**Comment by dnlib** [comment]
I have a proposed fix for you, I have tested it in my use-case and determined that it worked.

In Modules/IO/Embedded/EmbeddedResourceBuilder.cs I've changed line 62 from using `assembly.Location` to the following

```csharp
var sourceFile = new FileInfo(AppContext.BaseDirectory);
```

Give it a shot for yourselves and let me know!

**Comment by Kaliumhexacyanoferrat** [comment]
Thanks, I will have a look. I can not even build the app:

```
andreas.naegeli@NTB101 MSYS /c/Work/GenHTTP/GenHTTP/Playground (master)
$ dotnet publish -c Release -o publish -p:PublishReadyToRun=true -p:PublishSingleFile=true
```

yields

```
C:\Work\GenHTTP\GenHTTP\Modules\IO\Embedded\EmbeddedResourceContainer.cs(28,43): error IL3000: 'System.Reflection.Assembly.Location' always returns an empty string for assemblies embedded in a single-file app. If the path to the app directory is needed, consider calling 'System.AppContext.BaseDirectory'. [C:\Work\GenHTTP\GenHTTP\Modules\IO\GenHTTP.Modules.IO.csproj]
C:\Work\GenHTTP\GenHTTP\Modules\IO\Embedded\EmbeddedResourceBuilder.cs(62,43): error IL3000: 'System.Reflection.Assembly.Location' always returns an empty string for assemblies embedded in a single-file app. If the path to the app directory is needed, consider calling 'System.AppContext.BaseDirectory'. [C:\Work\GenHTTP\GenHTTP\Modules\IO\GenHTTP.Modules.IO.csproj]
```

**Comment by Kaliumhexacyanoferrat** [comment]
Fixed in [v6.3.2](https://github.com/Kaliumhexacyanoferrat/GenHTTP/releases/tag/v6.3.2), should be available via nuget anytime soon.

@dnlib: Just a note: your example adds the compression feature twice to the server (via `Defaults` and manually).

**Comment by dnlib** [comment]
Alright, I appreciate the fix and pointer!
